palace/linalg/errorestimator.cpp

```diff
@@ -0,0 +1,395 @@
+    const auto &coarse_fe = *coarse_flux_fes.GetFE(e);
+    auto &T = *coarse_flux_fes.GetElementTransformation(e);
+    mass_integrator.AssembleElementMatrix(coarse_fe, T, scalar_mass_matrices[e]);
+
+    const auto &smooth_fe = *smooth_flux_fes.GetFinestFESpace().GetFE(e);
+    coarse_fe.Project(smooth_fe, T, smooth_to_coarse_embed[e]);
+  }
+}
+
+IndicatorsAndNormalization CurlFluxErrorEstimator::operator()(const ComplexVector &v,
+                                                              bool normalize) const
+{
+  mfem::ParComplexGridFunction field(&fes);
+  field.real().SetFromTrueDofs(v.Real());
+  field.imag().SetFromTrueDofs(v.Imag());
+  const int nelem = smooth_flux_fes.GetFinestFESpace().GetNE();
+
+  // Coefficients for computing the discontinuous flux., i.e. (W, μ⁻¹∇ × V).
+  CurlFluxCoefficient real_coef(field.real(), mat_op), imag_coef(field.imag(), mat_op);
+  auto rhs_from_coef = [](mfem::ParFiniteElementSpace &flux_fes, auto &coef)
+  {
+    Vector RHS(flux_fes.GetTrueVSize());
+
+    mfem::LinearForm rhs(&flux_fes);
+    rhs.AddDomainIntegrator(new VectorFEDomainLFIntegrator(coef));
+    rhs.UseFastAssembly(false);
+    rhs.Assemble();
+    flux_fes.GetProlongationMatrix()->MultTranspose(rhs, RHS);
+
+    return RHS;
+  };
+
+  const auto smooth_flux_rhs =
+      ComplexVector(rhs_from_coef(smooth_flux_fes.GetFinestFESpace(), real_coef),
+                    rhs_from_coef(smooth_flux_fes.GetFinestFESpace(), imag_coef));
+
+  // Given the RHS vector of non-smooth flux, construct a flux projector and perform mass
+  // matrix inversion in the appropriate space, giving f = M⁻¹ f̂.
+  auto build_flux = [](const FluxProjector &proj, const ComplexVector &flux_coef)
+  {
+    // Use a copy construction to match appropriate size.
+    ComplexVector flux(flux_coef);
+    proj.Mult(flux_coef, flux);
+    return flux;
+  };
+  auto smooth_flux = build_flux(smooth_projector, smooth_flux_rhs);
+
+  // Given a complex solution represented with a PetscParVector, build a ComplexGridFunction
+  // for evaluation.
+  auto build_func = [](const ComplexVector &f, mfem::ParFiniteElementSpace &fes)
+  {
+    mfem::ParComplexGridFunction flux(&fes);
+    flux.real().SetFromTrueDofs(f.Real());
+    flux.imag().SetFromTrueDofs(f.Imag());
+    flux.real().ExchangeFaceNbrData();
+    flux.imag().ExchangeFaceNbrData();
+    return flux;
+  };
+  auto smooth_flux_func = build_func(smooth_flux, smooth_flux_fes.GetFinestFESpace());
+  mfem::ParComplexGridFunction coarse_flux_func(&coarse_flux_fes);
+  coarse_flux_func.real().ProjectCoefficient(real_coef);
+  coarse_flux_func.imag().ProjectCoefficient(imag_coef);
+
+  // Loop over elements, embed the smooth flux into the coarse flux space, then compute
+  // squared integral using a component-wise mass matrix.
+  Vector smooth_vec, coarse_vec, sub_vec, estimates(nelem);
+  estimates = 0.0;
+  double normalization = 0.0;
+  for (int e = 0; e < fes.GetNE(); e++)
```

Comment:
In thinking about the cost of allocating `scalar_mass_matrices` and `smooth_to_coarse_embed` above, I'm confused about this loop. Why is this not done matrix free by just looping over elements, looping over integration points on each element, and using `GridFunction::GetValue` or `GetVectorValue` to reconstruct the two quantities? This is essentially what is already happening when you call `coarse_flux_func.real().ProjectCoefficient(real_coef);` above, so you are doing double work here. See `GridFunction::ComputeElementLpErrors`.